gio/src/subclass/list_model.rs

```diff
@@ -0,0 +1,92 @@
+use crate::ListModel;
+use glib::subclass::prelude::*;
+use glib::translate::*;
+use glib::{Cast, IsA, ObjectExt};
+use once_cell::sync::Lazy;
+
+pub trait ListModelImpl: ObjectImpl {
+    fn get_item_type(&self, list_model: &Self::Type) -> glib::Type;
+    fn get_n_items(&self, list_model: &Self::Type) -> u32;
+    fn get_item(&self, list_model: &Self::Type, position: u32) -> Option<glib::Object>;
+}
+
+unsafe impl<T: ListModelImpl> IsImplementable<T> for ListModel
+where
+    <T as ObjectSubclass>::Type: IsA<glib::Object>,
+{
+    unsafe extern "C" fn interface_init(
+        iface: glib::ffi::gpointer,
+        _iface_data: glib::ffi::gpointer,
+    ) {
+        let list_model_iface = &mut *(iface as *mut ffi::GListModelInterface);
+
+        list_model_iface.get_item_type = Some(list_model_get_item_type::<T>);
+        list_model_iface.get_n_items = Some(list_model_get_n_items::<T>);
+        list_model_iface.get_item = Some(list_model_get_item::<T>);
+    }
+}
+
+static LIST_ITEM_TYPE_QUARK: Lazy<glib::Quark> =
+    Lazy::new(|| glib::Quark::from_string("gtk-rs-subclass-list-model-item-type"));
+
+unsafe extern "C" fn list_model_get_item_type<T: ListModelImpl>(
+    list_model: *mut ffi::GListModel,
+) -> glib::ffi::GType
+where
+    <T as ObjectSubclass>::Type: IsA<glib::Object>,
+{
+    let instance = &*(list_model as *mut T::Instance);
+    let imp = instance.get_impl();
+    let model = from_glib_borrow::<_, ListModel>(list_model);
```

Comment:
For consistency with the `IOStream` code, maybe also call it `wrap` here and elsewhere :)